### PR TITLE
Fix Clang compilation error with Lib ATen for ppc64le

### PR DESCRIPTION
**Pull request by pratiklp00**
This patch fixes error while compiling with Clang for ppc64le
I have used clang version 15.0.7 
Errors are as follow:
```
No matching function for call to 'vec_sel’
No matching function for call to 'vec_splats'
Excess elements in scalar initializer
Use of undeclared identifier 'vec_vsubudm'
Fix for multiple error within int64_t  DEFINE_MEMBER_OP_AND_ONE
```
References:
- https://releases.llvm.org/9.0.0/tools/clang/docs/AttributeReference.html
- https://reviews.llvm.org/D81083

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Comment by pratiklp00** [comment]
@malfet @XiaobingSuper Hi, please let me know if there is any suggestion or change/comment

**Review by malfet — CHANGES_REQUESTED** [review]
Few remakrs:
 - Please specify clang version in PR description and reference some doc explaining why `__attribute((altivec(vector__))` does not work for clang, but works for gcc , and vice versa (i.e. if say gcc-9+ supports `__attribute__((vector_size(16))` as alternative to altivec attribute, perhaps entire code can be changed to it
 - Please explain why `vint64` needs to be defined as `signed long long` rather than `signed long` or even `int64_t`? (This would also avoid force casting `vec_splats`, wouldn't it?)

File: aten/src/ATen/cpu/vec/vec256/vsx/vsx_helpers.h [hunk 1]
```diff
@@ -84,7 +100,13 @@ vec_sldw_aux(const vfloat32& vec_in0, const vfloat32& vec_in1) {
 #endif
 
 #define vec_not(a) vec_nor(a, a)
-
+#if defined(__clang__)
+#if !defined(vec_splats)
```
[the remark this answers is not in the document]
Comment (pratiklp00) [reply aten/src/ATen/cpu/vec/vec256/vsx/vsx_helpers.h:104]: No, I will combine it.

File: aten/src/ATen/cpu/vec/vec256/vsx/vec256_int64_vsx.h [hunk 2]
```diff
@@ -37,7 +38,7 @@ class Vectorized<int64_t> {
   C10_ALWAYS_INLINE Vectorized(vint64 v1, vint64 v2) : _vec0{v1}, _vec1{v2} {}
   C10_ALWAYS_INLINE Vectorized(vbool64 v1, vbool64 v2) : _vecb0{v1}, _vecb1{v2} {}
   C10_ALWAYS_INLINE Vectorized(int64_t scalar)
-      : _vec0{vec_splats(scalar)}, _vec1{vec_splats(scalar)} {}
+      : _vec0{vec_splats((ElementType)scalar)}, _vec1{(vint64)vec_splats((ElementType)scalar)} {}
```
[the remark this answers is not in the document]
Comment (pratiklp00) [reply aten/src/ATen/cpu/vec/vec256/vsx/vec256_int64_vsx.h:41]: I have removed them.

File: aten/src/ATen/cpu/vec/vec256/vsx/vsx_helpers.h [hunk 3]
```diff
@@ -84,7 +100,13 @@ vec_sldw_aux(const vfloat32& vec_in0, const vfloat32& vec_in1) {
...
+C10_ALWAYS_INLINE vint64 vec_splats(const int64_t& a) {
+ return vec_splats(static_cast<signed long long>(a));
```
[the remark this answers is not in the document]
Comment (pratiklp00) [reply aten/src/ATen/cpu/vec/vec256/vsx/vsx_helpers.h:106]: No it is not different. I will remove the casting.

**Comment by vinithakv** [comment]
@pratiklp00 
Some extra spaces reported by linter. Otherwise looks good to me.

**Comment by pratiklp00** [comment]
@malfet Hi, I have made the required changes Please review it.

**Comment by pratiklp00** [comment]
@malfet @ezyang @ezyang Hi Please let me know if any changes are required.

**Comment by malfet** [comment]
@pratiklp00 at the very list current form of the change break CI all over the place, can you please fix it?

File: aten/src/ATen/cpu/vec/vec256/vsx/vec256_complex_float_vsx.h [hunk 4]
```diff
@@ -146,8 +146,8 @@ class Vectorized<ComplexFlt> {
     auto mask_complex = Vectorized<ComplexFlt>(
         vec_mergeh(mask._vec0, mask._vec0), vec_mergeh(mask._vec1, mask._vec1));
     return {
-        vec_sel(a._vec0, b._vec0, mask_complex._vec0),
-        vec_sel(a._vec1, b._vec1, mask_complex._vec1),
+         vec_sel(a._vec0, b._vec0,reinterpret_cast<vbool32>(mask_complex._vec0)),       
+         vec_sel(a._vec1, b._vec1,reinterpret_cast<vbool32>(mask_complex._vec1)),
```
Comment (malfet) [inline aten/src/ATen/cpu/vec/vec256/vsx/vec256_complex_float_vsx.h:150]:
```suggestion
        vec_sel(a._vec0, b._vec0,reinterpret_cast<vbool32>(mask_complex._vec0)),       
        vec_sel(a._vec1, b._vec1,reinterpret_cast<vbool32>(mask_complex._vec1)),
```

File: aten/src/ATen/cpu/vec/vec256/vsx/vec256_complex_float_vsx.h [hunk 5]
```diff
@@ -156,8 +156,8 @@ class Vectorized<ComplexFlt> {
       const Vectorized<ComplexFlt>& b,
       const Vectorized<ComplexFlt>& mask) {
     return {
-        vec_sel(a._vec0, b._vec0, mask._vec0),
-        vec_sel(a._vec1, b._vec1, mask._vec1),
+         vec_sel(a._vec0, b._vec0, reinterpret_cast<vbool32>(mask._vec0)),
+         vec_sel(a._vec1, b._vec1, reinterpret_cast<vbool32>(mask._vec1)),
```
Comment (malfet) [inline aten/src/ATen/cpu/vec/vec256/vsx/vec256_complex_float_vsx.h:160]:
```suggestion
        vec_sel(a._vec0, b._vec0, reinterpret_cast<vbool32>(mask._vec0)),
        vec_sel(a._vec1, b._vec1, reinterpret_cast<vbool32>(mask._vec1)),
```

**Review by malfet — APPROVED** [review]

File: aten/src/ATen/cpu/vec/vec256/vsx/vec256_complex_float_vsx.h [hunk 6]
```diff
@@ -146,8 +146,8 @@ class Vectorized<ComplexFlt> {
     auto mask_complex = Vectorized<ComplexFlt>(
         vec_mergeh(mask._vec0, mask._vec0), vec_mergeh(mask._vec1, mask._vec1));
     return {
-        vec_sel(a._vec0, b._vec0, mask_complex._vec0),
-        vec_sel(a._vec1, b._vec1, mask_complex._vec1),
+        vec_sel(a._vec0, b._vec0,reinterpret_cast<vbool32>(mask_complex._vec0)),
```
Comment (malfet) [inline aten/src/ATen/cpu/vec/vec256/vsx/vec256_complex_float_vsx.h:149]:
```suggestion
        vec_sel(a._vec0, b._vec0,reinterpret_cast<vbool32>(mask_complex._vec0)),
```

File: aten/src/ATen/cpu/vec/vec256/vsx/vsx_helpers.h [hunk 7]
```diff
@@ -17,6 +33,7 @@ using vuint32  =  __attribute__((altivec(vector__)))  unsigned  int;
 using vuint64  =  __attribute__((altivec(vector__)))  unsigned long long;
 using vfloat32 =  __attribute__((altivec(vector__)))  float;
 using vfloat64 =  __attribute__((altivec(vector__)))  double;
+#endif
```
Comment (malfet) [inline aten/src/ATen/cpu/vec/vec256/vsx/vsx_helpers.h:36]:
```suggestion
#endif
```

File: aten/src/ATen/cpu/vec/vec256/vsx/vsx_helpers.h [hunk 8]
```diff
@@ -84,7 +100,11 @@ vec_sldw_aux(const vfloat32& vec_in0, const vfloat32& vec_in1) {
 #endif
 
 #define vec_not(a) vec_nor(a, a)
-
+#if defined(__clang__) && !defined(vec_splats)
+C10_ALWAYS_INLINE vint64 vec_splats(const int64_t& a) {
+  return vec_splats(a);
+ }
+#endif
```
Comment (malfet) [inline aten/src/ATen/cpu/vec/vec256/vsx/vsx_helpers.h:107]:
```suggestion
#endif
```

File: aten/src/ATen/cpu/vec/vec256/vsx/vec256_complex_float_vsx.h [hunk 9]
```diff
@@ -146,8 +146,8 @@ class Vectorized<ComplexFlt> {
...
+        vec_sel(a._vec1, b._vec1,reinterpret_cast<vbool32>(mask_complex._vec1)),
```
Comment (malfet) [inline aten/src/ATen/cpu/vec/vec256/vsx/vec256_complex_float_vsx.h:150]:
```suggestion
        vec_sel(a._vec0, b._vec0, reinterpret_cast<vbool32>(mask_complex._vec0)),
        vec_sel(a._vec1, b._vec1, reinterpret_cast<vbool32>(mask_complex._vec1)),
```

File: aten/src/ATen/cpu/vec/vec256/vsx/vsx_helpers.h [hunk 10]
```diff
@@ -84,7 +100,11 @@ vec_sldw_aux(const vfloat32& vec_in0, const vfloat32& vec_in1) {
...
```
Comment (malfet) [inline aten/src/ATen/cpu/vec/vec256/vsx/vsx_helpers.h:107]:
```suggestion
#if defined(__clang__) && !defined(vec_splats)
C10_ALWAYS_INLINE vint64 vec_splats(const int64_t& a) {
  return vec_splats(a);
}
#endif
```

**Pull request closed without merging**